### ISSUE TITLE
Remove bitcount field of the mpf tuple?

### ISSUE THREAD

**Issue by skirpichev** [mpmath/mpmath#720]
[Our docs](https://mpmath.readthedocs.io/en/latest/technical.html#representation-of-numbers) says:
> More technically, mpmath represents a floating-point number as a 4-tuple (sign, man, exp, bc) where sign is 0 or 1 (indicating positive vs negative) and the mantissa is nonnegative; bc (bitcount) is the size of the absolute value of the mantissa as measured in bits. Though redundant, keeping a separate sign field and explicitly keeping track of the bitcount significantly speeds up arithmetic (the bitcount, especially, is frequently needed but slow to compute from scratch due to the lack of a Python built-in function for the purpose).

Since CPython 3.1 there is a such function.  It's the [int.bit_length()](https://docs.python.org/3/library/stdtypes.html#int.bit_length) method.

Some timings:
```
$ python -m timeit -r11 -s 'from mpmath.libmp.libintmath import python_bitcount as bc' 'bc(1000)'
500000 loops, best of 11: 965 nsec per loop
$ python -m timeit -r11 -s 'from mpmath.libmp.libintmath import python_bitcount as bc' 'bc(10)'
500000 loops, best of 11: 930 nsec per loop
$ python -m timeit -r11 -s 'from mpmath.libmp.libintmath import python_bitcount as bc' -s 'from math import factorial;n=factorial(100)' 'bc(n)'
100000 loops, best of 11: 3.18 usec per loop
```
vs
```
$ python -m timeit -r11 '(1000).bit_length()'
5000000 loops, best of 11: 75.6 nsec per loop
$ python -m timeit -r11 '(10).bit_length()'
5000000 loops, best of 11: 74.9 nsec per loop
$ python -m timeit -r11 -s 'from math import factorial;n=factorial(100)' 'n.bit_length()'
5000000 loops, best of 11: 97.2 nsec per loop
```

So, using the bit_length() is clearly better and it's available for all backends.  Does it make sense to keep now the bc field of the mpf tuple or it should be deprecated?

**Comment by oscarbenjamin** [comment]
The tricky thing is that the mpf tuple is used as a conversion format in many places so there would still be a need to generate the 4-tuple with bit count often.

Otherwise though I agree that a 3 tuple is better and it can save a lot of wasted arithmetic (and bugs) trying to keep the bit count in sync.

**Comment by skirpichev** [comment]
> The tricky thing is that the mpf tuple is used as a conversion format in many places so there would still be a need to generate the 4-tuple with bit count often.

The proposed change is obviously not a backward compatible, so we will accept 4-tuples for a while, e.g. in ``_mpf_`` properties.

> Otherwise though I agree that a 3 tuple is better and it can save a lot of wasted arithmetic (and bugs)

But maybe we miss some scenarios where "caching" the bitcount has some benefits?  (Same for the sign bit, BTW.)

**Comment by skirpichev** [comment]
BTW, simpler format (2-tuple) was planned in #178.  So, I guess Fredrik is ok with this idea.

**Comment by skirpichev** [comment]
Currently, we use bc for special numbers (nan, inf, etc), but this seems redundant as well.

**Comment by skirpichev** [comment]
It seems, some mpmath users (well, the SymPy of course!) actually do tuple unpacking of mpf's.  I think we should deprecate this in the next release.  from_man_exp/to_man_exp() functions should be used instead.

**Comment by skirpichev** [comment]
Ok.  The plan is to deprecate direct manipulations with ``_mpf_``.  People should use ``from_man_exp()``/``to_man_exp()`` functions for finite numbers and dedicated imports for special forms like ``fnan``, ``fnzero``, ``finf`` and ``fninf``.  Then, after a release, we could change internal format of the ``_mpf_``.

Unfortunately, we can't emit a warning at runtime.

I'll make a patch for the Diofant to show how it's possible.  It seems, very few places are affected.  I think it's valid to the SymPy as well.

**Comment by skirpichev** [comment]
Ok, I guess, there are no fans of the bc field.  That's missing here is a warning in release notes for v1.4, that direct access to mpf tuple is deprecated.

**Comment by skirpichev** [comment]
@oscarbenjamin, beware that SymPy sometimes still access directly items of mpf tuples.  A quick grep:
```
$ git grep "_mpf_.*\["
sympy/core/numbers.py:    _mpf_: tuple[int, int, int, int]
sympy/core/numbers.py:        return x._mpf_[2] >= 0
$ git grep -E '(x|re|im|mag)\[[0-4]\]' sympy/core/evalf.py 
sympy/core/evalf.py:    "x[2] + (x[3] if x[1] != 1 else 0)".
sympy/core/evalf.py:    return x[2] + x[3]
sympy/core/evalf.py:        return (mag[0][0],) + mag[1:]
sympy/core/evalf.py:        if yre[0] == 1:  # y < 0
sympy/core/evalf.py:        return re[0], im, re[2], prec
```
Maybe there are more cases.

This was deprecated, see CHANGES.  Unfortunately, there is no way to issue a warning.

**Comment by oscarbenjamin** [comment]
I don't think it is a good idea to just deprecate the `_mpf_` attribute because it is used in many places as an interchange format including in python-flint and I assume SageMath and probably other code that depends on those or sympy.

If the internal representation used within mpmath changes to a 2-tuple or a 3-tuple then that seems reasonable to me but in that case the `_mpf_` attribute should be preserved as a property method that generates a 4-tuple for compatibility. Also there should be some new way to get the 2-tuple or 3-tuple rather than just deprecating `x._mpf_` or changing what it resolves to.

**Comment by skirpichev** [comment]
> I don't think it is a good idea to just deprecate the `_mpf_` attribute

No, it's not deprecated.  But please consider mpf tuples as opaque objects.  We have API to work with, discussed in https://github.com/mpmath/mpmath/issues/704#issuecomment-2953536980.  Let me know if something is missing (or not marked as public).

> as an interchange format including in python-flint

This is a backward-compatible patch for python-flint:
```diff
diff --git a/src/flint/types/arb.pyx b/src/flint/types/arb.pyx
index dbb5c06..7f84688 100644
--- a/src/flint/types/arb.pyx
+++ b/src/flint/types/arb.pyx
@@ -53,7 +53,7 @@ cdef arb_from_str(str s):
         raise ValueError("invalid string for arb()")
 
 cdef arb_set_mpmath_mpf(arb_t x, obj):
-    sgn, man, exp, _ = obj
+    sgn, man, exp, *_ = obj
 
     if not man:
         if not exp:
```

> but in that case the `_mpf_` attribute should be preserved as a property method that generates a 4-tuple for compatibility.

In the linked pr I just ignore last field in the mpf's constructor.  (Note, that at present `_mpf_` and `_mpmath_` attributes aren't documented at all.)  That mitigates compatibility break.

But I would like to avoid the path you suggest, if possible.  It essentially requires renaming of the `_mpf_` attribute, used internally for doing math, and provide a new property.  This also leak implementation details, e.g. encoding of special numbers.

And a new property will do useless computations just to keep compatibility.  This will be not a fastest way for conversion *from* mpmath.

BTW, the change should be benchmarked and maybe it worth to postpone it, until we do more severe changes (e.g. switch to using 2-tuple representation).

**Comment by oscarbenjamin** [comment]
What is the expected way to convert to/from a mpf and a 4-tuple?

> And a new property will do useless computations just to keep compatibility. This will be not a fastest way for conversion _from_ mpmath.

No, but it will preserve compatibility with other code. Code that uses `_mpf_` or that converts to/from using 4-tuples exists in a few places. There is no good reason to break that by changing what `_mpf_` means.

Note e.g.
```python
In [2]: flint.arb(3)._mpf_
Out[2]: (0, mpz(3), 0, 2)
```
Hence
```python
In [4]: mpmath.mpf(flint.arb(3))
Out[4]: mpf('3.0')
```
I'm sure that similar exists in SageMath and other places.

**Comment by skirpichev** [comment]
> What is the expected way to convert to/from a mpf and a 4-tuple?

Why you would like back 4-tuple?

If you want to convert from/to raw mpf tuple, there are from/to_man_exp() for finite numbers.  There are also finf, fninf and fnan constants with clear meaning.

> No, but it will preserve compatibility with other code.

At cost that conversion (*from* mpf) will be slower without warnings.

> Code that uses _mpf_ or that converts to/from using 4-tuples exists in a few places.

Good, less code to adapt.

> There is no good reason to break that by changing what _mpf_ means.

See issue description and https://github.com/mpmath/mpmath/issues/178.  There are reasons to change internals of the type, used in low-level API (libmp).

> Note e.g.
> ```
> In [2]: flint.arb(3)._mpf_
> Out[2]: (0, mpz(3), 0, 2)
> ```

As I said before, it's not a problem to preserve compatibility for conversions *to mpf*.

The problem is that you would like to use private `_mpf_` property of the mpf type, expecting certain its format.  That means, that after any change like removing bc field - this attribute can't be used by mpf internally to do arithmetics, using the libmp API.

**Comment by fredrik-johansson** [comment]
IMHO this seems like it will break compatibility with existing code for very little benefit. If it's really advantageous to change the internal format, then the 4-tuple `_mpf_` could still remain implemented as a property for compatibility.

**Comment by oscarbenjamin** [comment]
> > What is the expected way to convert to/from a mpf and a 4-tuple?
> 
> Why you would like back 4-tuple?

Because there is code that uses it:
```python
In [3]: flint.arb(mpmath.mpf(3))
Out[3]: 3.00000000000000
```

> The problem is that you would like to use private `_mpf_` property of the mpf type, expecting certain its format.

It is not private. It is a protocol that is used in various places:
https://github.com/flintlib/python-flint/blob/2e613be9a3a5e8cac1f1cc7518c9c3cec121819f/src/flint/types/arb.pyi#L16-L17

If you want to recommend new ways over using `_mpf_` then go ahead but there is no benefit in removing or changing the `_mpf_` attribute itself.

**Comment by skirpichev** [comment]
> If it's really advantageous to change the internal format

Lets first measure benefits of this.  I appreciate any suggestions for benchmarks.

But even with same performance we will have less code to track bc fields, IMO it's a good thing.

> then the 4-tuple _mpf_ could still remain implemented as a property for compatibility.

Yes, I sketched above this route:
1. rename `_mpf_` attribute of mpf to something else, say `_mpf`.
2. make `_mpf_` a property, that compute backward-compatible 4-tuple.

This introduce a lot of code churn in the `_mpf` class and a silent (unless we issue a warning) performance penalty for users of that `_mpf_` property.

> It is not private.

"_single_leading_underscore: weak “internal use” indicator." (c) PEP 8.

> It is a protocol that is used in various places

Where it's documented?

The bc field looks entirely redundant in this "protocol", anyway.  It's a leaked implementation detail, nothing more.

**Comment by oscarbenjamin** [comment]
> This introduce a lot of code churn in the `_mpf` class

This whole idea is code churn. There is a big difference though between churning the code that you maintain yourself vs forcing downstream churn and breakage. If you want to spend time churning the code then go ahead but changing the `_mpf_` attribute is just a pointless compatibility break. There are much more useful things that we can all spend our time doing than this like having a layer that can use python-flint's arf type rather than mpf tuples.

**Comment by skirpichev** [comment]
> This whole idea is code churn.

That was your opinion before:
> Otherwise though I agree that a 3 tuple is better and it can save a lot of wasted arithmetic (and bugs) trying to keep the bit count in sync.

> There are much more useful things that we can all spend our time doing than this

Do you believe it's not worth to develop the mp context or libmp layer?  Just bug fixes?

> like having a layer that can use python-flint's arf type rather than mpf tuples.

That should be possible, if this type get missing arithmetic methods (true division, for example) and all numbers.Real stuff.  Probably it's not hard to fix, it seems that the FLINT type has all required basic methods.  The real problem is lack of methods, implementing libmp's functions.  Perhaps, temporary conversions to arb could be used to implement things like sin().

There is a separate issue, https://github.com/mpmath/mpmath/issues/733, about this.

**Comment by oscarbenjamin** [comment]
> Do you believe it's not worth to develop the mp context or libmp layer? Just bug fixes?

I'm not sure what you mean by "develop" here. So far what I see are changes like renaming `mpf_log` to `mpf_ln` that don't really improve anything. What are the improvements you are thinking of for the libmp layer?

**Comment by oscarbenjamin** [comment]
> > like having a layer that can use python-flint's arf type rather than mpf tuples.
> 
> That should be possible, if this type get missing arithmetic methods (true division, for example) and all numbers.Real stuff. Probably it's not hard to fix, it seems that the FLINT type has all required basic methods. The real problem is lack of methods, implementing libmp's functions. Perhaps, temporary conversions to arb could be used to implement things like sin().

The main problem with using arf right now in mpmath is the fact that it uses a global precision variable making it not safe to use in multithreaded code or when mixed with any other code that manipulates python-flint's global context. What is needed is an alternate version of arf that has contexts with a context-local precision variable.

If that were fixed though then I would imagine that mpmath would use arf as an alternative to mpf tuples for basic arithmetic but could still implement e.g. sin over the top. The question is what is a good interface that makes it possible to swap out arf for mpf tuples transparently in something like the libmp code.

**Comment by skirpichev** [comment]
> So far what I see are changes like renaming mpf_log to mpf_ln that don't really improve anything.

If it's all you see in recent changes in mpmath, I think it would be better to keep myself from participation in the mpmath development.  Sorry, for breaking your code.

**Comment by oscarbenjamin** [comment]
Sorry, I don't mean to be negative about your contributions here. I just mean that I don't understand what is the bigger picture for how things in mpmath will develop and what that will bring in terms of meaningful improvements in the long term.

If there are going to be changed in libmp then what is the longer term plan for where it will end up and how will that be better?

**Comment by skirpichev** [comment]
> I just mean that I don't understand what is the bigger picture for how things in mpmath will develop and what that will bring in terms of meaningful improvements in the long term.

The rough plan was to make the mp context be close to the gmpy2 behavior for mpfr/mpc types.  That means, e.g. support for limited exponent range (with default to infinite or "big enough", as in the gmpy2), support for floating-point exceptions.  See next release milestone for more.

The libmp layer would be cut down to a minimum set of functions, announced in #704 as public.  The rest goes to mpmath/functions/.

I assumed, that (as it was announced) we can treat operands of libmp's functions as opaque objects and change internal structure of mpf's tuples to get performance bonus and/or simplify code.  Besides dropping the bc field, it could be using `(man, exp)` representation for finite numbers (if we will not support of the signed zero).

So, I though, it would be possible to gradually evolve the mp context *and* the libmp layer along this road.  But now I don't think so, if we had to maintain the mpf's tuples as part of the public API.  Then better to leave things as is and just fix bugs, where it's possible with the current design.

**Comment by skirpichev** [comment]
I benchmarked closed pr a bit:
```
$ export MPMATH_NOGMPY=Y
$ alias pp='python -m pyperf timeit -q --rigorous --copy-env'
$ pp -s 'from mpmath import mp,sin;mp.prec=100;z=1+1j' 'sin(z)'
Mean +- std dev: 87.3 us +- 2.9 us
$ pp -s 'from mpmath import mp,sin;mp.prec=300;z=1+1j' 'sin(z)'
Mean +- std dev: 118 us +- 1 us
$ pp -s 'from mpmath import mp,gamma;mp.prec=100;z=1+1j' 'gamma(z)'
Mean +- std dev: 388 us +- 6 us
$ git checkout no-bc/720
Switched to a new branch 'no-bc/720'
$ pp -s 'from mpmath import mp,sin;mp.prec=100;z=1+1j' 'sin(z)' 
Mean +- std dev: 82.8 us +- 0.7 us
$ pp -s 'from mpmath import mp,sin;mp.prec=300;z=1+1j' 'sin(z)' 
Mean +- std dev: 113 us +- 1 us
$ pp -s 'from mpmath import mp,gamma;mp.prec=100;z=1+1j' 'gamma(z)'
Mean +- std dev: 374 us +- 4 us
```
(Similar happens with gmpy2/gmp, using adapted helpers.)

So, in principle there is measurable speed gain ~4-5%.  Maybe better with a polished version of the patch.
